### PR TITLE
Change logical operator for relationship filters targeting interfaces

### LINKED ISSUE #4704: Incorrect Logical Operators used by relationship filters.
**Describe the bug**
The query:
```graphql
{
    actors(
        where: {
            actedInConnection_ALL: { node: { actorsConnection_ALL: { node: { name: "Keanu Reeves" } } } }
        }
    ) {
        actedIn {
            title
        }
    }
}
```
Produce the wrong results when `actedInConnection_ALL` is targeting an interface. 

**Type definitions**
```graphql

interface Show {
    title: String!
    actors: [Actor!]! @declareRelationship
}

type Movie implements Show {
    title: String!
    cost: Float
    runtime: Int
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN, properties: "ActedIn")
}

type Series implements Show {
    title: String!
    episodes: Int
    actors: [Actor!]! @relationship(type: "STARRED_IN", direction: IN, properties: "StarredIn")
}

type Actor {
    name: String!
    actedIn: [Show!]! @relationship(type: "ACTED_IN", direction: OUT, properties: "ActedIn")
}

type ActedIn @relationshipProperties {
    screenTime: Int
}

type StarredIn @relationshipProperties {
    episodes: Int
}
```

**To Reproduce**

Run the following query:
```graphql
{
    actors(
        where: {
            actedInConnection_ALL: { node: { actorsConnection_ALL: { node: { name: "Keanu Reeves" } } } }
        }
    ) {
        actedIn {
            title
        }
    }
}
```
The Cypher produced is:
```graphql
MATCH (this:Actor)
WHERE ((EXISTS {
    MATCH (this)-[this0:ACTED_IN]->(this1:Movie)
    WHERE (EXISTS {
        MATCH (this1)<-[this2:ACTED_IN]-(this3:Actor)
        WHERE this3.name = $param0
    } AND NOT (EXISTS {
        MATCH (this1)<-[this2:ACTED_IN]-(this3:Actor)
        WHERE NOT (this3.name = $param0)
    }))
} AND NOT (EXISTS {
    MATCH (this)-[this0:ACTED_IN]->(this1:Movie)
    WHERE NOT (EXISTS {
        MATCH (this1)<-[this2:ACTED_IN]-(this3:Actor)
        WHERE this3.name = $param0
    } AND NOT (EXISTS {
        MATCH (this1)<-[this2:ACTED_IN]-(this3:Actor)
        WHERE NOT (this3.name = $param0)
    }))
})) OR (EXISTS {
    MATCH (this)-[this4:ACTED_IN]->(this5:Series)
    WHERE (EXISTS {
        MATCH (this5)<-[this6:STARRED_IN]-(this7:Actor)
        WHERE this7.name = $param1
    } AND NOT (EXISTS {
        MATCH (this5)<-[this6:STARRED_IN]-(this7:Actor)
        WHERE NOT (this7.name = $param1)
    }))
} AND NOT (EXISTS {
    MATCH (this)-[this4:ACTED_IN]->(this5:Series)
    WHERE NOT (EXISTS {
        MATCH (this5)<-[this6:STARRED_IN]-(this7:Actor)
        WHERE this7.name = $param1
    } AND NOT (EXISTS {
        MATCH (this5)<-[this6:STARRED_IN]-(this7:Actor)
        WHERE NOT (this7.name = $param1)
    }))
})))
CALL {
    WITH this
    CALL {
        WITH *
        MATCH (this)-[this8:ACTED_IN]->(this9:Movie)
        WITH this9 { .title, __resolveType: \\"Movie\\", __id: id(this9) } AS this9
        RETURN this9 AS var10
        UNION
        WITH *
        MATCH (this)-[this11:ACTED_IN]->(this12:Series)
        WITH this12 { .title, __resolveType: \\"Series\\", __id: id(this12) } AS this12
        RETURN this12 AS var10
    }
    WITH var10
    RETURN collect(var10) AS var10
}
RETURN this { actedIn: var10 } AS this
```
As shown above, the approach used by the library is to generate every concrete entity an `EXISTS` subquery for each concrete entity implemented by the interface "Show". However, the logical operator used under the hood remains independent from the filter used.
This will produce incorrect data, as for the `_SOME` operator it will make sense that at least one implementation fulfills the predicate, and therefore seems correct to use the Logical Operator `OR`, but for the operator `_ALL` the expected behavior is that the predicate should be fulfilled by each concrete entity.

**Expected behavior**
The logical operator used should reflect the operation used:
- `ALL` ->  should be true across all the concrete entities
- `SOME` -> should be true for some concrete entites
- `SINGLE` ->  should be true exactly once
- `NONE` -> should be true across all the concrete entites

**System (please complete the following information):**

* OS: [e.g. macOS]
* Version: [@neo4j/graphql@4.5]
* Node.js version: [18.0.0]

### PR DESCRIPTION
**Pull request by MacondoExpress**
# Description

This PR changes the Logical Operator used when multiple relationship filters are generated when the target of the relationship is an interface, as described by the issue: https://github.com/neo4j/graphql/issues/4704

Complexity: Low

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Thanks for the Neo4j GraphQL Toolbox updates.

The Neo4j GraphQL Toolbox has now been torn down - reopening this PR will republish it.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 35 | 45 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 43 | 65 | 1134 |
|  aggregations.NestedAggregation | 16544 | 16544 | 54 | 79 | 2174 |
|  aggregations.AggregationWithWhere | 11970 | 11970 | 36 | 61 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116478 | 22116478 | 2159 | 2259 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116478 | 22116478 | 2260 | 2226 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9701 | 9701 | 30 | 45 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9928 | 9928 | 31 | 37 | 2174 |
|  aggregations.DeeplyNestedCount | 11062184 | 11062184 | 2866 | 2995 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 42 | 61 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 74 | 151 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 122 | 146 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 92 | 132 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 140 | 184 | 3003 |
|  connections.Connection | 14049 | 14049 | 65 | 139 | 2174 |
|  connections.NestedConnection | 41068 | 41068 | 127 | 143 | 4516 |
|  create.SimpleMutation | 7 | 7 | 50 | 60 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 41 | 31 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 220 | 202 | 1040 |
|  delete.NestedDeleteInUpdate | 19165 | 19165 | 173 | 203 | 1183 |
|  2871.NestedRelationshipFilter | 23127 | 23009 | 74 | 58 | 4443 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 37 | 43 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22891 | 22891 | 74 | 87 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 37 | 38 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 49 | 61 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 20 | 22 | 1040 |
|  query.SimpleQueryWithRelationship | 16129 | 16129 | 37 | 51 | 2174 |
|  query.QueryWhere | 9694 | 9693 | 30 | 40 | 2167 |
|  query.SimpleQueryWithNestedWhere | 9861 | 9860 | 44 | 50 | 2167 |
|  query.Nested | 10095502 | 10095502 | 7097 | 7019 | 2008534 |
|  query.NestedWithFilter | 10074351 | 10074351 | 6678 | 6659 | 2004000 |
|  query.OrFilterOnRelationships | 41610 | 42103 | 263 | 175 | 1479 |
|  query.OrFilterOnRelationshipsAndNested | 35297 | 36223 | 185 | 265 | 1479 |
|  query.QueryWithNestedIn | 13761 | 13358 | 51 | 56 | 1345 |
|  query.NestedConnectionWhere | 9801 | 9801 | 53 | 68 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9991 | 9866 | 79 | 88 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18932 | 18746 | 126 | 155 | 1646 |
|  query.NestedWithRelationshipSingleFilters | 3865 | 3865 | 151 | 200 | 1134 |
|  query.Fulltext | 64 | 64 | 28 | 37 | 16 |
|  query.FulltextWithNestedQuery | 558 | 558 | 41 | 54 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14059 | 14059 | 45 | 52 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14059 | 14059 | 69 | 107 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14170 | 14159 | 69 | 71 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14170 | 14159 | 116 | 110 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14583 | 14574 | 151 | 153 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14402 | 14500 | 135 | 163 | 2174 |
|  sorting.SortMultipleTypes | 3498 | 3498 | 102 | 109 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14476 | 14400 | 125 | 134 | 2174 |
|  sorting.SortOnNestedFields | 14049 | 14049 | 48 | 54 | 2174 |
|  sorting.SortDeeplyNestedFields | 43148 | 43148 | 84 | 95 | 4516 |
|  sorting.ConnectionWithSort | 3281 | 3281 | 74 | 94 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 58 | 69 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 55 | 61 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 250 | 293 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 254 | 279 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 111 | 118 | 2002 |

</details>
        
Old Schema Generation: 26.322s
Schema Generation: 27.357s
Old Subgraph Schema Generation: 29.780s
Subgraph Schema Generation: 30.587s

File: packages/graphql/src/translate/queryAST/ast/filters/LogicalFilter.ts [hunk 1]
```diff
@@ -64,6 +64,9 @@ export class LogicalFilter extends Filter {
             case "OR": {
                 return Cypher.or(...predicates);
             }
+            case "XOR": {
+                return Cypher.xor(...predicates);
+            }
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/filters/LogicalFilter.ts:69]: do we support this in graphql?

File: packages/graphql/src/translate/queryAST/factory/parsers/parse-where-field.ts [hunk 2]
```diff
@@ -52,6 +52,9 @@ export function parseWhereField(field: string): WhereRegexGroups {
             operator = notSplit[1] as WhereOperator;
         } else if (matchGroups.operator === "NOT" || matchGroups.operator === "NONE") {
             isNot = true;
+            if (matchGroups.operator === "NONE") { 
+                operator = notSplit[0] as WhereOperator;
+            }
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/factory/parsers/parse-where-field.ts:57]: Isn't there a bit of duplicate logic here by having the `isNot` as well as the operator set to none?

File: packages/graphql/tests/tck/experimental/interface-relationship-authorization.test.ts [hunk 3]
```diff
@@ -253,4 +253,58 @@ describe("Interface filtering on authorization rules", () => {
...
+            }
+        `;
+
+        const token = createBearerToken("secret", {});
+        const result = await translateQuery(neoSchema, query, { token });
+
+        expect(formatCypher(result.cypher)).toMatchInlineSnapshot(`
+            "MATCH (this:Actor)
+            WITH *
+            WHERE apoc.util.validatePredicate(NOT ($isAuthenticated = true AND (single(this1 IN [(this)-[this3:ACTED_IN]->(this1:Movie) WHERE single(this0 IN [(this1)<-[this2:ACTED_IN]-(this0:Actor) WHERE ($param1 IS NOT NULL AND this0.name = $param1) | 1] WHERE true) | 1] WHERE true) XOR single(this5 IN [(this)-[this7:ACTED_IN]->(this5:Series) WHERE single(this4 IN [(this5)<-[this6:STARRED_IN]-(this4:Actor) WHERE ($param2 IS NOT NULL AND this4.name = $param2) | 1] WHERE true) | 1] WHERE true))), \\"@neo4j/graphql/FORBIDDEN\\", [0])
```
Comment (angrykoala) [inline packages/graphql/tests/tck/experimental/interface-relationship-authorization.test.ts:298]: I'm not sure I understand why we need `XOR` here :sweat:

**Review by angrykoala — APPROVED** [review]

File: packages/graphql/tests/tck/experimental/interface-relationship-authorization.test.ts (hunk 3 above)
Comment (MacondoExpress) [reply packages/graphql/tests/tck/experimental/interface-relationship-authorization.test.ts:298]: This is required for the `_SINGLE` operator as by generating the same `EXISTS` subquery filter for each concrete entity, the same filter could be true by more than one implementation, using the `XOR` will produce exactly one filter across all the concrete entities fulfill the predicate.

File: packages/graphql/src/translate/queryAST/ast/filters/LogicalFilter.ts (hunk 1 above)
Comment (MacondoExpress) [reply packages/graphql/src/translate/queryAST/ast/filters/LogicalFilter.ts:69]: Replied below!

File: packages/graphql/src/translate/queryAST/factory/parsers/parse-where-field.ts (hunk 2 above)
Comment (MacondoExpress) [reply packages/graphql/src/translate/queryAST/factory/parsers/parse-where-field.ts:57]: Unfortunately not, it's a very long rambling story, but the gist is that we convert the syntax for `not operator` like `actedIn: {}` defined as `_SOME`, `_NONE` as `_NOT(_SOME))`  etc... A mix of premature optimization and several syntaxes supported.

**Pull request merged**